Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp

```diff
@@ -0,0 +1,145 @@
+    m_clickMeasurement = WTFMove(attribution);
+    completionHandler();
+}
+
+void EphemeralStore::markAllUnattributedPrivateClickMeasurementAsExpiredForTesting()
+{
+    if (m_clickMeasurement && !m_clickMeasurement->attributionTriggerData())
+        reset();
+}
+
+void EphemeralStore::attributePrivateClickMeasurement(WebCore::PCM::SourceSite&& sourceSite, WebCore::PCM::AttributionDestinationSite&& destinationSite, const ApplicationBundleIdentifier& applicationBundleIdentifier, WebCore::PCM::AttributionTriggerData&& attributionTriggerData, WebCore::PrivateClickMeasurement::IsRunningLayoutTest isLayoutTest, CompletionHandler<void(std::optional<WebCore::PCM::AttributionSecondsUntilSendData>&&, DebugInfo&&)>&& completionHandler)
+{
+    if (!m_clickMeasurement)
+        return completionHandler(std::nullopt, { });
+
+    if (m_clickMeasurement->sourceSite() != sourceSite || m_clickMeasurement->destinationSite() != destinationSite)
+        return completionHandler(std::nullopt, { });
+
+    if (!applicationBundleIdentifier.isEmpty() && m_clickMeasurement->sourceApplicationBundleID() != applicationBundleIdentifier)
+        return completionHandler(std::nullopt, { });
+
+    completionHandler(m_clickMeasurement->attributeAndGetEarliestTimeToSend(WTFMove(attributionTriggerData), isLayoutTest), { });
+}
+
+void EphemeralStore::privateClickMeasurementToStringForTesting(CompletionHandler<void(String)>&& completionHandler) const
+{
+    if (!m_clickMeasurement)
+        return completionHandler("\nNo stored Private Click Measurement data.\n"_s);
+
+    StringBuilder builder;
+    builder.append("\nEphemeral Private Click Measurement:\n"_s);
+    builder.append("SourceSite: "_s, m_clickMeasurement->sourceSite().registrableDomain.string(), "\n"_s);
+    builder.append("DestinationSite: "_s, m_clickMeasurement->destinationSite().registrableDomain.string(), "\n"_s);
+    builder.append("SourceID: "_s, m_clickMeasurement->sourceID(), "\n"_s);
+    if (auto trigger = m_clickMeasurement->attributionTriggerData()) {
+        builder.append("Trigger data: "_s, trigger->data, "\n"_s);
+        builder.append("Trigger priority: "_s, trigger->priority, "\n"_s);
+    }
+    return completionHandler(builder.toString());
+}
+
+void EphemeralStore::allAttributedPrivateClickMeasurement(CompletionHandler<void(Vector<WebCore::PrivateClickMeasurement>&&)>&& completionHandler)
+{
+    if (m_clickMeasurement && m_clickMeasurement->attributionTriggerData())
+        completionHandler({ *m_clickMeasurement });
+    else
+        completionHandler({ });
+}
+
+void EphemeralStore::markAttributedPrivateClickMeasurementsAsExpiredForTesting(CompletionHandler<void()>&& completionHandler)
+{
+    if (m_clickMeasurement && m_clickMeasurement->attributionTriggerData())
+        reset();
+    completionHandler();
+}
+
+void EphemeralStore::clearPrivateClickMeasurement(CompletionHandler<void()>&& completionHandler)
```

Comment:
Since ephemeral and persistent sessions can exist in parallel, please make sure that higher level calls to clear/reset PCM does so for both the ephemeral and the persistent store.

Reply:
This comment made me realize that `clearPrivateClickMeasurement` _should_ probably be implemented even for the ephemeral data store, if only for robustness in the future. In practice, it doesn't really matter since we have the maxAge check *before* measuring and attributing the click anyways, but it's nicer to have parity with the persistent version.